gordon/router.py

```diff
@@ -121,16 +123,44 @@ def _get_next_phase(self, event_msg):
             next_phase = 'cleanup'
         return next_phase
 
+    async def _update_messages_in_flight(self, event_msg):
+        if event_msg.msg_id not in self._messages_in_flight:
+            context = {'unit': 'seconds'}
+            timer = self.metrics.timer(
+                'router-message-flight-duration', context=context)
+            await timer.start()
+            self._messages_in_flight[event_msg.msg_id] = timer
+
+        elif event_msg.phase in self.FINAL_PHASES:
+            timer = self._messages_in_flight.pop(event_msg.msg_id)
+            await timer.stop()
+
+        in_flight_count = len(self._messages_in_flight)
+        await self.metrics.set(
+            'router-messages-in-flight', value=in_flight_count)
+
     async def _route(self, event_msg, next_phase=None):
         if not interfaces.IEventMessage.providedBy(event_msg):
             msg = (f'Ignoring message "{event_msg}". Does not correctly '
                    'implement `IEventMessage`.')
             logging.warn(msg)
+
+            context = {'error': 'invalid-message-provider'}
+            await self.metrics.incr('router-message-dropped', context=context)
             return
 
+        await self._update_messages_in_flight(event_msg)
+
         if not next_phase:
             next_phase = self._get_next_phase(event_msg)
 
+        old_phase = event_msg.phase
+        context = {
+            'current': old_phase,
+            'next': next_phase,
+        }
+        await self.metrics.incr('router-update-message-phase', context=context)
```

Comment:
Ah I missed that one in the commit - good catch. This is to see message by phase. @descent3me and I were talking through what to measure last week, and I originally had this idea for each plugin, i.e. "enricher initially receives message" and "publisher initially receives message" etc, so we can track the difference between consume -> publish and consume -> enrich -> publish. But then @descent3me noticed that this can just live in the router since it knows all this info already.

Reply:
(updated the PR message, will update the actual long-desc of commit message too)